src/nupic/algorithms/SpatialPooler.cpp

```diff
@@ -863,14 +867,20 @@ void SpatialPooler::bumpUpWeakColumns_() {
 
 
 void SpatialPooler::updateDutyCyclesHelper_(vector<Real> &dutyCycles,
-                                            const vector<UInt> &newValues,
+                                            SDR &newValues,
                                             UInt period) {
   NTA_ASSERT(period > 0);
-  NTA_ASSERT(dutyCycles.size() == newValues.size());
+  NTA_ASSERT(dutyCycles.size() == newValues.size);
 
-  for (Size i = 0; i < dutyCycles.size(); i++) {
-    dutyCycles[i] = (dutyCycles[i] * (period - 1) + newValues[i]) / period;
-  }
+  newValues.print();
+
+  const Real decay = (Real) (period - 1) / period;
+  for (Size i = 0; i < dutyCycles.size(); i++)
+    dutyCycles[i] *= decay;
+
+  const Real increment = 1. / period;
+  for(const auto &idx : newValues.getFlatSparse())
+    dutyCycles[idx] += increment;
```

Comment:
nit, might add explanation (I got stucked here for a while) 
`// dutyCycles[i] += newValues[i]/period`